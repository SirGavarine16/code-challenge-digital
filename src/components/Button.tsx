import React, { FC } from 'react';
import { FlexStyle, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import { useTheme } from '@react-navigation/native';

interface Props {
    title?: string;
    onPress?: () => void;
    width?: FlexStyle['width'];
}

const Button: FC<Props> = ({ title = '', onPress, width = '100%' }) => {
    const { colors } = useTheme();
    const { container, text } = styles;

    return (
        <TouchableOpacity activeOpacity={0.8} testID='button' onPress={onPress} style={[container, { backgroundColor: colors.primary, width }]}>
            <Text testID='button-title' style={text}>
                {title}
            </Text>
        </TouchableOpacity>
    );
};

const styles = StyleSheet.create({
    container: {
        padding: 15,
        borderRadius: 10,
    },
    text: {
        color: 'white',
        fontSize: 14,
        fontWeight: '800',
        textAlign: 'center',
    }
})

export default Button;
