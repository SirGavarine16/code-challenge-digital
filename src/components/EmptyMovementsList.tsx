import React, { FC } from 'react';
import { StyleSheet, Text } from 'react-native';

import { colors } from '../themes/AppTheme';

interface Props {
    message?: string;
}

const EmptyMovementsList: FC<Props> = ({ message = 'No hay movimientos disponibles.' }) => {
    return (
        <Text testID='empty-message' style={styles.text}>
            {message}
        </Text>
    );
};

const styles = StyleSheet.create({
    text: {
        fontSize: 16,
        fontWeight: '400',
        textAlign: 'center',
    }
});

export default EmptyMovementsList;
