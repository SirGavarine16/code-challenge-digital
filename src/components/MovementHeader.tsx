import React, { FC } from 'react';
import { StyleSheet, Text, View } from 'react-native';

import { colors, spacing } from '../themes/AppTheme';

interface Props {
    product?: string;
    top?: number;
}

const MovementHeader: FC<Props> = ({ product = 'Nombre del producto', top = 0 }) => {
    const { container, text } = styles;

    return (
        <View style={[container, { paddingTop: top + 80 }]}>
            <Text testID='product-name' style={text}>
                {product}
            </Text>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        backgroundColor: colors.secondary,
        paddingHorizontal: spacing.paddingHorizontal,
        marginBottom: 20,
    },
    text: {
        fontSize: 24,
        fontWeight: '800',
        marginBottom: 10,
    }
})

export default MovementHeader;
