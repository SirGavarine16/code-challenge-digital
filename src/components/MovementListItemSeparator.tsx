import React, { FC } from 'react';
import { StyleSheet, View } from 'react-native';

interface Props {
}

const MovementListItemSeparator: FC<Props> = () => {
    const { container } = styles;
    return (
        <View style={container} />
    );
};

const styles = StyleSheet.create({
    container: {
        height: 5,
    },
});

export default MovementListItemSeparator;
