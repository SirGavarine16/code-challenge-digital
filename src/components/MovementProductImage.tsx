import React, { FC } from 'react';
import { Image, StyleSheet, View } from 'react-native';

import { colors } from '../themes/AppTheme';

interface Props {
    image: string;
}

const MovementProductImage: FC<Props> = ({ image }) => {
    const { container, content } = styles;

    return (
        <View testID='card-image' style={container}>
            <Image testID='product-image' source={{ uri: image }}  style={content} resizeMode='contain' />
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        backgroundColor: 'white',
        width: '100%',
        aspectRatio: 1 / 1,
        shadowColor: "#000",
        shadowOffset: {
            width: 0,
            height: 4,
        },
        shadowOpacity: 0.30,
        shadowRadius: 4.65,
        elevation: 8,
        borderRadius: 20,
        justifyContent: 'center',
        alignItems: 'center',
    },
    content: {
        width: '100%',
        height: '100%',
        backgroundColor: colors.background
    }
})

export default MovementProductImage;
