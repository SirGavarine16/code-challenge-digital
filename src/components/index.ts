export { default as WelcomeHeader } from './WelcomeHeader';
export { default as SectionLabel } from './SectionLabel';
export { default as PointsDisplay } from './PointsDisplay';
export { default as MovementsList } from './MovementsList';
export { default as MovementCard } from './MovementCard';
export { default as MovementListItemSeparator } from './MovementListItemSeparator';
export { default as Button } from './Button';
export { default as MovementHeader } from './MovementHeader';
export { default as MovementProductImage } from './MovementProductImage';
export { default as EmptyMovementsList } from './EmptyMovementsList';