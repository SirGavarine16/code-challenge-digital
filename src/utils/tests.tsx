import { PropsWithChildren } from 'react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { SafeAreaProvider } from 'react-native-safe-area-context';

import { rootReducer } from '../redux/store';

export const SafeAreaWrapper = ({ children }: PropsWithChildren<{}>) => {
    return <SafeAreaProvider>{children}</SafeAreaProvider>
}

export const ReduxWrapper = ({ children }: PropsWithChildren<{}>) => {
    const store = configureStore({ reducer: rootReducer });
    return <Provider store={store}>{children}</Provider>
}