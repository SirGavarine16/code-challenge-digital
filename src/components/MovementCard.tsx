import React, { FC } from 'react';
import { Image, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';

import { Movement } from '../interfaces';
import { formatDate } from '../utils/formats';
import { AppStackParams } from '../navigation/AppStack';

interface Props {
    item: Movement;
}

const MovementCard: FC<Props> = ({ item }) => {
    const { container, image, dataContainer, boldText, normalText, pointsContainer, pointsText } = styles;

    const { navigate } = useNavigation<StackNavigationProp<AppStackParams>>();

    return (
        <TouchableOpacity testID='card' style={container} onPress={() => navigate('MovementScreen', item)}>
            <Image testID='image' source={{ uri: item.image }} style={image} resizeMode='contain' />
            <View style={dataContainer}>
                <Text testID='name' style={boldText}>
                    {item.product}
                </Text>
                <Text testID='date' style={[normalText, { marginTop: 2 }]}>
                    {formatDate(item.createdAt)}
                </Text>
            </View>
            <View style={pointsContainer}>
                <MaterialCommunityIcons testID='points-icon' name={item.is_redemption ? 'minus' : 'plus'} color={item.is_redemption ? '#FF0000' : '#00B833'} />
                <Text testID='points' style={pointsText}>
                    {item.points}
                </Text>
            </View>
            <MaterialCommunityIcons size={25} name='chevron-right' color='#070707'  />
        </TouchableOpacity>
    );
};

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
        overflow: 'hidden',
    },
    image: {
        width: 55,
        height: 55,
        resizeMode: 'contain',
        borderRadius: 5,
    },
    dataContainer: {
        marginLeft: 10,
        maxWidth: '50%',
        justifyContent: 'center'
    },
    normalText: {
        fontSize: 12,
        fontWeight: '400',
    },
    boldText: {
        fontWeight: '800'
    },
    pointsContainer: {
        flex: 1,
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'flex-end'
    },
    pointsText: {
        fontSize: 16,
        fontWeight: '800',
    }
});

export default MovementCard;
