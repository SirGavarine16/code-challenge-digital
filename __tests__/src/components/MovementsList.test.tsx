import { render } from '@testing-library/react-native';

import { MovementsList } from '../../../src/components';
import { ReduxWrapper } from '../../../src/utils/tests';

describe('<MovementsList />', () => {
    const renderComponent = () => render(<MovementsList />, { wrapper: ReduxWrapper });

    it('should render correctly', () => {
        const screen = renderComponent();
        expect(screen).toMatchSnapshot();
    });

    it('should display a section label', () => {
        const { getByTestId } = renderComponent();
        expect(getByTestId('section-label')).toBeDefined();
    });

    it('should display a list container', () => {
        const { getByTestId } = renderComponent();
        expect(getByTestId('list-container')).toBeDefined();
    });

    it('secton label should be Tus movimientos', () => {
        const { getByTestId } = renderComponent();
        expect(getByTestId('section-label').children[0]).toBe('Tus movimientos');
    });
});
