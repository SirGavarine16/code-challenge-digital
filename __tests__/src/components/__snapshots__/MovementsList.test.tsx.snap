// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovementsList /> should render correctly 1`] = `
<View
  style={
    [
      {
        "flex": 1,
      },
      {
        "paddingBottom": 20,
      },
    ]
  }
>
  <Text
    style={
      {
        "color": "#9B9898",
        "fontSize": 14,
        "fontWeight": "800",
        "marginBottom": 20,
        "textTransform": "uppercase",
      }
    }
    testID="section-label"
  >
    Tus movimientos
  </Text>
  <View
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 8,
        "flex": 1,
        "padding": 10,
      }
    }
    testID="list-container"
  >
    <RCTScrollView
      ItemSeparatorComponent={[Function]}
      ListEmptyComponent={[Function]}
      data={[]}
      getItem={[Function]}
      getItemCount={[Function]}
      getItemLayout={[Function]}
      keyExtractor={[Function]}
      maxToRenderPerBatch={8}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onRefresh={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
        />
      }
      refreshing={false}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <RCTRefreshControl />
      <View>
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "400",
              "textAlign": "center",
            }
          }
          testID="empty-message"
        >
          No hay movimientos disponibles.
        </Text>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 30,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "rgb(0, 122, 255)",
          "borderRadius": 10,
          "opacity": 1,
          "padding": 15,
          "width": "48.5%",
        }
      }
      testID="button"
    >
      <Text
        style={
          {
            "color": "white",
            "fontSize": 14,
            "fontWeight": "800",
            "textAlign": "center",
          }
        }
        testID="button-title"
      >
        Ganados
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "rgb(0, 122, 255)",
          "borderRadius": 10,
          "opacity": 1,
          "padding": 15,
          "width": "48.5%",
        }
      }
      testID="button"
    >
      <Text
        style={
          {
            "color": "white",
            "fontSize": 14,
            "fontWeight": "800",
            "textAlign": "center",
          }
        }
        testID="button-title"
      >
        Canjeados
      </Text>
    </View>
  </View>
</View>
`;
